tests/linen/module_test.py

```diff
@@ -615,9 +648,37 @@ def __call__(self, x):
     self.assertEqual(trace, expected_trace)
 
 
+  def test_module_apply_method(self):
+    class Foo(nn.Module):
+      @nn.compact
+      def __call__(self):
+        pass
+      
+      def test(self):
+        pass
+
+    # We can use both instance and class methods in apply.
+    Foo().apply({}, method=Foo.test)
+    Foo().apply({}, method=Foo().test)
+
+    # We can even use a function that is not in the provided Module, although
+    # it requires an argument for the class (Foo in this case).
```

Comment:
I don't understand this -- the lambda would get passed `cls` as the first argument? Is this a real use-case? What would someone do with `cls`? (I guess I don't understand how you'd use this pattern)

Reply:
This example is a bit contrived, but I wanted to test that the first argument to the custom apply function is actually the `Foo` class. I did that by directly returning it with a lambda and then checking the types of the returned value is indeed `Foo`.

I am not sure what you question is, but I think it is quite useful that someone can define a custom apply function without having to change the original Module:.

```python
class Foo(nn.Module):
  features: int = 4

  def setup(self):
    self.dense = nn.Dense(self.features)

  @nn.compact
  def __call__(self, x):
    x = self.dense(x)
    # do some stuff
    return x

def custom_apply(cls, x):
  x = cls.dense(x)
  # do some custom stuff
  return x

Foo().apply(..., ) # this calls Foo's apply
Foo().apply(..., method=custom_apply)  # this uses the custom apply function
```

Reply:
How does the line `x = cls.dense(x)` work?

Wouldn't it need to be an instance of Foo (with `self.features`?).

Reply:
No, I don't think that is how it works (but please correct me if I am wrong).  I think you don't need any instances in the apply method, you only need the classes, and apply will then instantiate everything automatically.

In `apply`, we first "unbound" the `__call__` function, which means it is no longer attached to `self`: https://github.com/google/flax/blob/fa826152a5c092c8f4e8a50d01f0d560ea7a6af3/flax/linen/module.py#L303

So in that sense `__call__` is exactly the same as any function that has a first argument representing the class. After creating an unbound instance of the function, the function is called with an instantiated clone of the module (`Foo` in this case), where the parent scope is initialized correctly.